### PR TITLE
[#1741] delete jwtsecret when deleting applicatie in the admin

### LINKED ISSUE #1741: Deleting an Application via the admin via a bulk-delete doesn't delete the underlying Autorisatie
### Product versie / Product version

1.12.3 and 1.11 (test.openzaak.nl)

### Omschrijf het probleem / Describe the bug

Taiga Utrecht 264

Having trailing entities when deleting an Application was an issue earlier in #1515 , this issue is new but related.

When deleting an indivudal Application things go well, however when you do a bulk-delete with the Application the Autorisatie (or JWTSecret?) object remains. Even though the Application credentials can no longer be used, the client-id's of the Application's Autorisatie can't be reused.

This issue can be easily reproduced by doing a bulk-delete of an Application and then attempting to use the deleted client-id again in a new Application:

image

image

### Stappen om te reproduceren / Steps to reproduce

_No response_

### Verwacht gedrag / Expected behavior

_No response_

### PR DESCRIPTION
**Pull request by annashamray**
Fixes #1741

### CODE REVIEW

File: src/openzaak/components/autorisaties/admin.py [hunk 1]
```diff
@@ -117,6 +117,17 @@ def delete_model(self, request, obj):
         secrets.delete()
         super().delete_model(request, obj)
 
+    @transaction.atomic
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        client_ids = queryset.values_list("client_ids", flat=True)
+        client_ids = [client_id[0] for client_id in client_ids]
```
Comment (Coperh) [inline src/openzaak/components/autorisaties/admin.py:124]:
```suggestion
        client_ids = [client_id for client_id  in client_id_field for client_id_field  in client_ids]
```
Change to something like this


This also apparently works. but I've never used it.
```
import itertools
a = [['a','b'], ['c']]
print(list(itertools.chain.from_iterable(a)))
```

Comment (annashamray) [reply src/openzaak/components/autorisaties/admin.py]: Thanks for catching it!
I've used `sum`, `itertools.chain` looks also interesting

**Review by Coperh — APPROVED** [review]
Yep, looks good

File: src/openzaak/components/autorisaties/tests/admin/test_applications_admin.py [hunk 2]
```diff
@@ -135,3 +135,32 @@ def test_create_applicatie_inline_credentials_without_secret(self):
...
+
+        # confirmation page
+        form = response.form
+        form["action"] = "delete_selected"
+        form["post"] = "yes"
+
+        response = form.submit()
+
+        self.assertEqual(response.status_code, 302)
+        self.assertFalse(JWTSecret.objects.exists())
```
Comment (stevenbal) [inline src/openzaak/components/autorisaties/tests/admin/test_applications_admin.py:166]: You could maybe add a JWTSecret that you don't expect to be deleted, to ensure the `delete_queryset` doesn't accidentally delete all JWTSecrets

File: src/openzaak/components/autorisaties/admin.py [hunk 3]
```diff
@@ -117,6 +117,17 @@ def delete_model(self, request, obj):
         secrets.delete()
         super().delete_model(request, obj)
 
+    @transaction.atomic
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        client_ids = queryset.values_list("client_ids", flat=True)
+        client_ids = sum(list(client_ids), [])
```
Comment (stevenbal) [inline src/openzaak/components/autorisaties/admin.py:124]: Is this line necessary?

**Review by stevenbal — CHANGES_REQUESTED** [review]

File: src/openzaak/components/autorisaties/admin.py (hunk 3 above)
Comment (annashamray) [reply src/openzaak/components/autorisaties/admin.py:124]: Yes, it flats the list
for example, if we delete 2 applicaties, which have client_ids: ['id1', id2'] and ['id3'], then the line123 will give the result:
client_ids = [ ['id1', id2'], ['id3'] ], so we need to flat it further

File: src/openzaak/components/autorisaties/tests/admin/test_applications_admin.py (hunk 2 above)
Comment (annashamray) [reply src/openzaak/components/autorisaties/tests/admin/test_applications_admin.py]: Sure, I've updated the test

**Comment by Coperh** [comment]
I just remembered, but do we want a management command to delete orphaned JWT secrets?

**Comment by annashamray** [comment]
@Coperh hmmm do we need a management command or a one-time data migration? If deleting jwt secrets work as intended there won't be any orphans anymore, right? 
But there is always a change of some bug of course...

**Comment by stevenbal** [comment]
@annashamray I'd probably add a data migration for it, otherwise people will have to run the management command manually

**Comment by annashamray** [comment]
@stevenbal @Coperh I've added a data migration to remove orphaned jwt secrets

**Comment by stevenbal** [comment]
@annashamray looks good :+1:

**Pull request merged**